keps/sig-cloud-provider/20191004-out-of-tree-credential-providers.md

```diff
@@ -0,0 +1,216 @@
+
+### Goals
+
+* Develop/test/release an interface for kubelet to obtain registry credentials from a cloud provider specific binary
+* Update/test/release the credential acquisition logic within kubelet
+* Build user documentation for out-of-tree credential providers
+* Support migration from existing in-tree credential providers to the new credential provider interface, along with dynamic roll back.
+* Remove in-tree credential provider code from kuberentes core
+
+### Non-Goals
+
+* Broad removal of cloud SDK usage falls under the [KEP for removing in-tree providers](https://github.com/kubernetes/enhancements/blob/master/keps/sig-cloud-provider/2019-01-25-removing-in-tree-providers.md).
+* Continuing to support projects that import the credential provider package.
+
+## Proposal
+
+### External Credential Provider
+
+An executable capable of providing container registry credentials will be pre-installed on each node so that it exists when kubelet starts running. This binary will be executed by the kubelet to obtain container registry credentials in a format compatible with container runtimes. Credential responses may be cached within the kubelet.
+
+This architecture is similar to the approach taken by the exec based credential plugin architecture already present in client-go and CNI, and is a well understood pattern.  The API types are modeled after the ExecConfig and ExecCredential in client-go which define exec based credential retrieval for similar use cases.
+
+A RegistryCredentialProvider configuration API type (similar to [clientauthentication](https://github.com/kubernetes/kubernetes/tree/0273d43ae9486e9d0be292c01de2dd4143522b86/staging/src/k8s.io/client-go/pkg/apis/clientauthentication/v1beta1)) will be added to Kubernetes:
+
+```go
+// RegistryCredentialProvider is used by the kubelet container runtime to match image URLs with
+// exec-based credential provider plugins that provide container registry credentials.
+type RegistryCredentialProvider struct {
+    metav1.TypeMeta `json:",inline"`
+
+    // ImageURLMatchers is a list of strings used to match against the image URL
+    // to determine which images to provide credentials for.  If one of the
+    // string matches the image URL, then the RegistryCredentialProvider will be
+    // used by kubelet to provide credentials for the image pull.
+    //
+    // Each ImageURLMatcher string is a URL pattern which can optionally contain
+    // a port and a path.  Globs can be used in the hostname (but not the port or
```

Comment:
- are these actually URLs or are they container specs (e.g. `k8s.gcr.io/foo` vs `https://k8s.gcr.io/foo`)?
- are globs only supported as suffixes, or also internally (e.g. `k8s.gcr.io/foo/*` vs `k8s.gcr.io/foo/*/bar`)?
- are globs only supported at path boundaries, or also for partial paths  (e.g. `k8s.gcr.io/foo/*` vs `k8s.gcr.io/foo*`)?
- are multiple providers supported or would that be delegated to an external "muxing" provider?

Reply:
1. Container specs
2. Internally
3. No, k8s*.gcr.io works
4. Single subdomain segment only
5. Multiple

I was going to base the matching logic off the existing urlsMatch function: https://github.com/kubernetes/kubernetes/blob/master/pkg/credentialprovider/keyring.go#L203

Reply:
That's fair.  According to [docker docs](https://docs.docker.com/engine/reference/commandline/pull/#pull-from-a-different-registry): 
> A registry path is similar to a URL, but does not contain a protocol specifier (https://).

Reply:
Ok, mostly for my own understanding: the docs references I found for the image field are `"The image property of a container supports the same syntax as the docker command does, including private registries and tags."` [1](https://kubernetes.io/docs/concepts/containers/images/) and `"The image property of a container supports the same syntax as the docker command does, including private registries and tags."` [2](https://docs.docker.com/engine/reference/commandline/pull/#pull-from-a-different-registry)

Reply:
> Please update the docs to reflect these answers.

Done.